core/environment.js

```diff
@@ -0,0 +1,352 @@
+    },
+
+    isAndroidDevice: {
+        get: function () {
+            if (this._isAndroidDevice === null) {
+                this._isAndroidDevice = this.platformName === Environment.ANDROID;
+            }
+
+            return this._isAndroidDevice;
+        }
+    },
+
+    isCordova: {
+        value: !!window.cordova
+    },
+
+    _isMobileDevice: {
+        value: null
+    },
+
+    isMobileDevice: {
+        get: function () {
+            if (this._isMobileDevice === null) {
+                // Android 4.4+ Chrome, Opera mobile 12+, Firefox 11.0+, Safari, IE (no tablet)
+                this._isMobileDevice = /mobile/gi.test(this.userAgent);
+            }
+
+            return this._isMobileDevice;
+        }
+    },
+
+    isAndroidTablet: {
+        get: function () {
+            return this.isAndroidDevice && !this.isMobileDevice;
+        }
+    },
+
+    _isStandalone: {
+        value: null
+    },
+
+    isStandalone: {
+        get: function () {
+            if (this._isStandalone === null) {
+                this._isStandalone = !!('standalone' in navigator && navigator.standalone);
+            }
+
+            return this._isStandalone;
+        }
+    },
+
+    _platformName: {
+        value: null
+    },
+
+    platformName: {
+        get: function () {
+            if (this._platformName === null) {
+                var match = this.userAgent.match(/android|windows\sphone|windows|macintosh|linux|cros/gi);
+
+                if (match && match.length) {
+                    var firstCandidate = match[0];
+
+                    if (firstCandidate === Environment.LINUX && match[1] === Environment.ANDROID) {
+                        firstCandidate = Environment.ANDROID;
+
+                    } else if (firstCandidate === CROS) {
+                        firstCandidate = Environment.CHROME_OS;
+                    }
+
+                    this._platformName = firstCandidate;
+
+                } else if (this.isIOSDevice) {
+                    this._platformName = Environment.IOS;
+
+                } else {
+                    this._platformName = UNKNOWN;
+                }
+            }
+
+            return this._platformName;
+        }
+    },
+
+    _browserVersion: {
+        value: null
+    },
+
+    browserVersion: {
+        get: function () {
+            if (this._browserVersion === null) {
+                this._analyzeBrowser();
+            }
+
+            return this._browserVersion;
+        }
+    },
+
+    _browserName: {
+        value: null
+    },
+
+    browserName: {
+        get: function () {
+            if (this._browserName === null) {
+                this._analyzeBrowser();
+            }
+
+            return this._browserName;
+        }
+    },
+
+    _isWKWebView: {
+        value: null
+    },
+
+    isWKWebView: {
+        get: function () {
+            if (this._isWKWebView === null) {
+                this._isWKWebView = this.isIOSDevice && !!window.indexedDB;
```

Comment:
This is penalizing every read, better to do:

return this._isWKWebView || (this._isWKWebView = this.isIOSDevice && !!window.indexedDB);

This applies to all methods here as much as possible